### ISSUE TITLE
POTENTIAL DEADLOCK: pkg/bpf/map.go:369 bpf.(*Map).OpenOrCreate

### ISSUE THREAD

**Issue by aanm** [cilium/cilium#4843]
```
2018-07-13T13:38:45.670397498Z POTENTIAL DEADLOCK: Inconsistent locking. saw this ordering in one goroutine:
2018-07-13T13:38:45.670423451Z happened before
2018-07-13T13:38:45.670653364Z /go/src/github.com/cilium/cilium/pkg/bpf/map.go:369 bpf.(*Map).OpenOrCreate ??? <<<<<
2018-07-13T13:38:45.670791437Z /go/src/github.com/cilium/cilium/daemon/daemon.go:851 main.(*Daemon).init ???
2018-07-13T13:38:45.670797023Z /go/src/github.com/cilium/cilium/daemon/daemon.go:1265 main.NewDaemon ???
2018-07-13T13:38:45.670800219Z /go/src/github.com/cilium/cilium/daemon/main.go:747 main.runDaemon ???
2018-07-13T13:38:45.670804221Z /go/src/github.com/cilium/cilium/daemon/main.go:131 main.glob..func1 ???
2018-07-13T13:38:45.670807682Z /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:648 cobra.(*Command).execute ???
2018-07-13T13:38:45.670811171Z /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:735 cobra.(*Command).ExecuteC ???
2018-07-13T13:38:45.670814405Z /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:693 cobra.(*Command).Execute ???
2018-07-13T13:38:45.670818201Z /go/src/github.com/cilium/cilium/daemon/main.go:145 main.main ???
2018-07-13T13:38:45.670831969Z /usr/local/go/src/runtime/proc.go:198 runtime.main ???
2018-07-13T13:38:45.670844494Z 
2018-07-13T13:38:45.670848697Z happened after
2018-07-13T13:38:45.670876119Z /go/src/github.com/cilium/cilium/pkg/bpf/map_register.go:30 bpf.registerMap ??? <<<<<
2018-07-13T13:38:45.670882212Z /go/src/github.com/cilium/cilium/pkg/bpf/map.go:401 bpf.(*Map).OpenOrCreate ???
2018-07-13T13:38:45.670886087Z /go/src/github.com/cilium/cilium/daemon/daemon.go:851 main.(*Daemon).init ???
2018-07-13T13:38:45.670889643Z /go/src/github.com/cilium/cilium/daemon/daemon.go:1265 main.NewDaemon ???
2018-07-13T13:38:45.670892502Z /go/src/github.com/cilium/cilium/daemon/main.go:747 main.runDaemon ???
2018-07-13T13:38:45.670895199Z /go/src/github.com/cilium/cilium/daemon/main.go:131 main.glob..func1 ???
2018-07-13T13:38:45.670898183Z /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:648 cobra.(*Command).execute ???
2018-07-13T13:38:45.670901324Z /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:735 cobra.(*Command).ExecuteC ???
2018-07-13T13:38:45.670904298Z /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:693 cobra.(*Command).Execute ???
2018-07-13T13:38:45.670921898Z /go/src/github.com/cilium/cilium/daemon/main.go:145 main.main ???
2018-07-13T13:38:45.670924893Z /usr/local/go/src/runtime/proc.go:198 runtime.main ???
2018-07-13T13:38:45.670927979Z 
2018-07-13T13:38:45.670930867Z in another goroutine: happened before
2018-07-13T13:38:45.670933869Z /go/src/github.com/cilium/cilium/pkg/bpf/map_register.go:62 bpf.GetOpenMaps ??? <<<<<
2018-07-13T13:38:45.67093724Z /go/src/github.com/cilium/cilium/daemon/map.go:52 main.(*getMap).Handle ???
2018-07-13T13:38:45.6709405Z /go/src/github.com/cilium/cilium/api/v1/server/restapi/daemon/get_map.go:54 daemon.(*GetMap).ServeHTTP ???
2018-07-13T13:38:45.670943464Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/operation.go:28 middleware.NewOperationExecutor.func1 ???
2018-07-13T13:38:45.670946522Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.670949557Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/router.go:73 middleware.NewRouter.func1 ???
2018-07-13T13:38:45.670952819Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.670955656Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/redoc.go:72 middleware.Redoc.func1 ???
2018-07-13T13:38:45.670958909Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.670961866Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/spec.go:45 middleware.Spec.func1 ???
2018-07-13T13:38:45.670965078Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.67098717Z /go/src/github.com/cilium/cilium/pkg/metrics/middleware.go:36 metrics.(*APIEventTSHelper).ServeHTTP ???
2018-07-13T13:38:45.670990684Z /go/src/github.com/cilium/cilium/pkg/api/apipanic.go:46 api.(*APIPanicHandler).ServeHTTP ???
2018-07-13T13:38:45.670993797Z /usr/local/go/src/net/http/server.go:2694 http.serverHandler.ServeHTTP ???
2018-07-13T13:38:45.670996936Z /usr/local/go/src/net/http/server.go:1830 http.(*conn).serve ???
2018-07-13T13:38:45.670999986Z 
2018-07-13T13:38:45.671004048Z happend after
2018-07-13T13:38:45.671007177Z /go/src/github.com/cilium/cilium/pkg/bpf/map.go:747 bpf.(*Map).GetModel ??? <<<<<
2018-07-13T13:38:45.671010392Z /go/src/github.com/cilium/cilium/pkg/bpf/map_register.go:67 bpf.GetOpenMaps ???
2018-07-13T13:38:45.671013502Z /go/src/github.com/cilium/cilium/daemon/map.go:52 main.(*getMap).Handle ???
2018-07-13T13:38:45.671016637Z /go/src/github.com/cilium/cilium/api/v1/server/restapi/daemon/get_map.go:54 daemon.(*GetMap).ServeHTTP ???
2018-07-13T13:38:45.6710199Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/operation.go:28 middleware.NewOperationExecutor.func1 ???
2018-07-13T13:38:45.671023232Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.671026223Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/router.go:73 middleware.NewRouter.func1 ???
2018-07-13T13:38:45.67102951Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.671032881Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/redoc.go:72 middleware.Redoc.func1 ???
2018-07-13T13:38:45.671036074Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.671039111Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/spec.go:45 middleware.Spec.func1 ???
2018-07-13T13:38:45.671042236Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.671045221Z /go/src/github.com/cilium/cilium/pkg/metrics/middleware.go:36 metrics.(*APIEventTSHelper).ServeHTTP ???
2018-07-13T13:38:45.67104812Z /go/src/github.com/cilium/cilium/pkg/api/apipanic.go:46 api.(*APIPanicHandler).ServeHTTP ???
2018-07-13T13:38:45.671051835Z /usr/local/go/src/net/http/server.go:2694 http.serverHandler.ServeHTTP ???
2018-07-13T13:38:45.671055235Z /usr/local/go/src/net/http/server.go:1830 http.(*conn).serve ???
2018-07-13T13:38:45.671058583Z 
2018-07-13T13:38:45.671061645Z 
2018-07-13T13:38:45.671064678Z Other goroutines holding locks:
2018-07-13T13:38:45.671067676Z goroutine 49241 lock 0x3dc6880
2018-07-13T13:38:45.671070889Z /go/src/github.com/cilium/cilium/pkg/bpf/map_register.go:62 bpf.GetOpenMaps ??? <<<<<
2018-07-13T13:38:45.671075081Z /go/src/github.com/cilium/cilium/daemon/map.go:52 main.(*getMap).Handle ???
2018-07-13T13:38:45.671079744Z /go/src/github.com/cilium/cilium/api/v1/server/restapi/daemon/get_map.go:54 daemon.(*GetMap).ServeHTTP ???
2018-07-13T13:38:45.671087515Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/operation.go:28 middleware.NewOperationExecutor.func1 ???
2018-07-13T13:38:45.671092013Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.671095941Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/router.go:73 middleware.NewRouter.func1 ???
2018-07-13T13:38:45.671099687Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.671103277Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/redoc.go:72 middleware.Redoc.func1 ???
2018-07-13T13:38:45.671106981Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.671110862Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/spec.go:45 middleware.Spec.func1 ???
2018-07-13T13:38:45.671114117Z /usr/local/go/src/net/http/server.go:1947 http.HandlerFunc.ServeHTTP ???
2018-07-13T13:38:45.671117765Z /go/src/github.com/cilium/cilium/pkg/metrics/middleware.go:36 metrics.(*APIEventTSHelper).ServeHTTP ???
2018-07-13T13:38:45.671121614Z /go/src/github.com/cilium/cilium/pkg/api/apipanic.go:46 api.(*APIPanicHandler).ServeHTTP ???
2018-07-13T13:38:45.671125215Z /usr/local/go/src/net/http/server.go:2694 http.serverHandler.ServeHTTP ???
2018-07-13T13:38:45.671128465Z /usr/local/go/src/net/http/server.go:1830 http.(*conn).serve ???
2018-07-13T13:38:45.671131997Z 
2018-07-13T13:38:45.671135712Z 
2018-07-13T13:38:45.671139423Z 
2018-07-13T13:38:45.675664731Z level=fatal msg="Agent pipe unexpectedly closed, shutting down"
```

**Comment by tgraf** [comment]
Cause was   `func GetOpenMaps() []*models.BPFMap {`  acquiring the per map mutex nested inside the global mutex.

**Comment by ianvernon** [comment]
Fixed-by #4854
